Update c2c documentation with Openshift info (#9861)
ballerina-platform/ballerina-dev-website: swan-lake/development-tutorials/deployment-guide/containerized-deployment.md

1. Docker
2. Kubernetes

> 3. Openshift

Reviewer: ```suggestion
3. OpenShift
```

>**Note:** You can visit the <a href="https://github.com/ballerina-platform/ballerina-spec/blob/master/c2c/code-to-cloud-spec.md" target="_blank">Code to Cloud specification</a> for detailed information about all the supported features. As mentioned in the beginning, Code to Cloud is a tool created to make the development process easier. It does not cover the operational properties in Kubernetes. If you want to have these additional features in your deployment, or you need to change the generated Kubernetes artifacts, you can use <a href="https://kustomize.io" target="_blank">Kustomize</a> for modifying the generated YAML.

## Openshift deployment

### Create the Ballerina package

> Below sample describes a Ballerina application that reads a greeting string from a config map and greets the user upon HTTP request. By following the steps below, you can make a Kubernetes deployment that has container resource limits, horizontal pod autoscaling, config maps and liveness, readiness probes with the help of Code to Cloud.

Reviewer: ```suggestion
The sample below describes a Ballerina application that reads a greeting string from a config map and greets the user upon an HTTP request. By following the steps below, you can make a Kubernetes deployment that has container resource limits, horizontal pod autoscaling, config maps, and liveness and readiness probes with the help of Code to Cloud.
```

2. Replace the content of the `main.bal` file with the content below.

   ***main.bal***

    ```ballerina
    import ballerina/http;

    listener http:Listener helloEP = new (9090);

    configurable string greeting = ?;

    service /helloWorld on helloEP {
        resource function get sayHello() returns string|error? {
            return greeting + ", Openshift!";
        }
    }
    ```
2. Create `probes.bal` with the following contents.

   ***probes.bal***

    ```ballerina
    import ballerina/http;

    listener http:Listener probeEP = new (9091);
    
    service /probes on probeEP {
        resource function get healthz() returns boolean {
            return true;
        }
        resource function get readyz() returns boolean {
            return true;
        }
    }
    ```
3. Create the `Config.toml` file in the package directory and paste the following content into it.

   ***Config.toml***
    ```toml
    greeting = "Hello"
    ```

4. Add `cloud = "openshift"` under the `[build-options]` table into the `Ballerina.toml` file in the package.

   ***Ballerina.toml***

    ```toml
    [build-options]
    cloud = "openshift"
    ```

> 5. Create a file named `Cloud.toml` in the package directory and add the content below. You can use some properties from the docker sample here since `cloud="k8s"` option builds both Docker image and Kubernetes. You can see a brief description of the properties in the comments.

Reviewer: ```suggestion
5. Create a file named `Cloud.toml` in the package directory and add the content below. You can use some properties from the Docker sample here since the `cloud="k8s"` option builds both Docker image and Kubernetes. You can see a brief description of the properties in the comments.
```

   ***Cloud.toml***

    ```toml
    [container.image]
    repository="wso2inc" # Ex - Docker hub repository name.
    name="hello-k8s" # Container name
    tag="v0.1.0"

    [cloud.deployment]
    min_memory="100Mi" # Minimum memory required for the container.
    max_memory="256Mi" # Maximum memory a single container can take.
    min_cpu="500m"  # Minimum CPU required for the container.
    max_cpu="500m" # Maximum CPU a single container can take.

    [cloud.deployment.autoscaling]
    min_replicas=2 # Minimum number of container instances ran at a given time
    max_replicas=5 # Maximum number of replicas of the container can scale up to
    cpu=60 # Average CPU utilization of pods

    [[cloud.config.files]] # Mounts the `Config.toml` file as a config map in Kubernetes.
    file="./Config.toml"

    [cloud.deployment.probes.liveness]
    port=9091
    path="/probes/healthz"

    [cloud.deployment.probes.readiness]
    port=9091
    path="/probes/readyz"
    ```

### Generate the artifacts

Once you build the Ballerina package, the compiler extension will generate the artifacts required for openshift deployment.

> >**Tip:** If you don't have the cloud="openshift" entry on the `Ballerina.toml` you can execute `bal build --cloud=openshift` command to provide the build option inline.

Reviewer: ```suggestion
>**Tip:** If you don't have the cloud="openshift" entry on the `Ballerina.toml`, you can execute the `bal build --cloud=openshift` command to provide the build option inline.
```

```
$ bal build
Compiling source
        wso2/hello_openshift:0.1.0

Generating executable

Generating artifacts...

        @kubernetes:Service                      - complete 1/2
        @kubernetes:Service                      - complete 2/2
        @kubernetes:ConfigMap                    - complete 1/1
        @kubernetes:Deployment                   - complete 1/1
        @kubernetes:HPA                          - complete 1/1
        @kubernetes:Docker                       - complete 2/2 

        Execute the below command to deploy the Kubernetes artifacts: 
        oc apply -f /home/wso2/c2c-guide/hello_openshift/target/openshift/hello_openshift

        target/bin/hello_openshift.jar
```

> >**Note:** Before invoking the Kubernetes service, observe the Openshift yaml that have been generated. You should be able to find services for ports that have been exposed, `HorizontalPodAutoscaler` for scaling and Deployment, and Config maps.

Reviewer: ```suggestion
>**Note:** Before invoking the Kubernetes service, observe the OpenShift YAML that has been generated. You should be able to find services for ports that have been exposed, `HorizontalPodAutoscaler` for scaling and Deployment, and ConfigMaps.
```

          initialDelaySeconds: 30
        name: "hello-openshift-deployment"
        ports:
        - containerPort: 9090
          name: "port-1-hello-openshift"
          protocol: "TCP"
        - containerPort: 9091
          name: "port-2-hello-openshift"
          protocol: "TCP"
        readinessProbe:
          httpGet:
            path: "/probes/readyz"
            port: 9091
          initialDelaySeconds: 30
        resources:
          limits:
            memory: "256Mi"
            cpu: "500m"
          requests:
            memory: "100Mi"
            cpu: "500m"
        volumeMounts:
        - mountPath: "/home/ballerina/conf/"
          name: "config-config-map-volume"
          readOnly: false
      nodeSelector: {}
      volumes:
      - configMap:
          name: "config-config-map"
        name: "config-config-map-volume"
---
apiVersion: "autoscaling/v1"
kind: "HorizontalPodAutoscaler"
metadata:
  labels:
    app: "hello_openshift"
  name: "hello-openshift-hpa"
spec:
  maxReplicas: 5
  minReplicas: 2
  scaleTargetRef:
    apiVersion: "apps/v1"
    kind: "Deployment"
    name: "hello-openshift-deployment"
  targetCPUUtilizationPercentage: 60
```

> ### Execute the Openshift service

Reviewer: ```suggestion
### Execute the Openshift service
```
